File: slack_bot/config/loader.py
"""Bot configuration loader. Reads from YAML files (bootstrap) or database."""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path

import psycopg2
import yaml


@dataclass
class PersonaConfig:
    display_name: str = ""
    description: str = ""
    personality: str = ""
    response_prefix: str = ""


@dataclass
class ProjectConfig:
    repo_path: str = ""
    github_repo: str = ""
    github_token_var: str = ""


@dataclass
class AirflowConfig:
    dag_prefix: str = ""


@dataclass
class ClaudeConfig:
    provider: str = "claude"  # "claude" or "codex"
    model: str = "sonnet"
    max_turns: int = 10
    full_agent: bool = False


@dataclass
class MemoryConfig:
    context_window: int = 20
    auto_extract: bool = True


@dataclass
class SecurityConfig:
    allowed_channels: list[str] = field(default_factory=list)
    allowed_users: list[str] = field(default_factory=list)
    dangerous_tools: list[str] = field(default_factory=list)
    mention_only: bool = False


@dataclass
class MattermostConfig:
    url: str = ""
    token: str = ""
    port: int = 8065


@dataclass
class DiscordConfig:
    token: str = ""
    guild_id: str = ""


@dataclass
class BotConfig:
    id: str
    name: str
    slack_app_token: str
    slack_bot_token: str
    platform: str = "slack"
    mattermost: MattermostConfig = field(default_factory=MattermostConfig)
    discord: DiscordConfig = field(default_factory=DiscordConfig)
    channels: list[str] = field(default_factory=list)
    persona: PersonaConfig = field(default_factory=PersonaConfig)
    project: ProjectConfig = field(default_factory=ProjectConfig)
    airflow: AirflowConfig = field(default_factory=AirflowConfig)
    tools_enabled: list[str] = field(default_factory=list)
    claude: ClaudeConfig = field(default_factory=ClaudeConfig)
    memory: MemoryConfig = field(default_factory=MemoryConfig)
    security: SecurityConfig = field(default_factory=SecurityConfig)


def _resolve_env(value: str) -> str:
    """Resolve ${ENV_VAR} references in string values."""
    if isinstance(value, str) and value.startswith("${") and value.endswith("}"):
        env_key = value[2:-1]
        return os.environ.get(env_key, "")
    return value


def _validate_bot_config(config: BotConfig) -> None:
    """Validate platform-specific required fields.

    Raises:
        ValueError: If required fields for the chosen platform are missing.
    """
    if config.platform == "slack":
        missing = [
            field
            for field, value in [
                ("slack_app_token", config.slack_app_token),
                ("slack_bot_token", config.slack_bot_token),
            ]
            if not value
        ]
        if missing:
            raise ValueError(
                f"Bot '{config.id}' (platform=slack) missing required fields: "
                + ", ".join(missing)
            )
    elif config.platform == "mattermost":
        missing = [
            field
            for field, value in [
                ("mattermost.url", config.mattermost.url),
                ("mattermost.token", config.mattermost.token),
            ]
            if not value
        ]
        if missing:
            raise ValueError(
                f"Bot '{config.id}' (platform=mattermost) missing required fields: "
                + ", ".join(missing)
            )
    elif config.platform == "discord":
        missing = [
            field
            for field, value in [
                ("discord.token", config.discord.token),
            ]
            if not value
        ]
        if missing:
            raise ValueError(
                f"Bot '{config.id}' (platform=discord) missing required fields: "
                + ", ".join(missing)
            )
    else:
        raise ValueError(
            f"Bot '{config.id}' has unsupported platform: '{config.platform}'. "
            "Supported values: 'slack', 'mattermost', 'discord'."
        )


def load_bot_from_yaml(path: Path) -> BotConfig:
    """Load a single bot configuration from a YAML file."""
    with open(path) as f:
        data = yaml.safe_load(f)

    slack = data.get("slack", {})
    mattermost_data = data.get("mattermost", {})
    discord_data = data.get("discord", {})
    persona_data = data.get("persona", {})
    project_data = data.get("project", {})
    airflow_data = data.get("airflow", {})
    claude_data = data.get("claude", {})
    memory_data = data.get("memory", {})
    security_data = data.get("security", {})
    tools_data = data.get("tools", {})

    config = BotConfig(
        id=data["name"],
        name=data["name"],
        platform=data.get("platform", "slack"),
        slack_app_token=_resolve_env(slack.get("app_token", "")),
        slack_bot_token=_resolve_env(slack.get("bot_token", "")),
        mattermost=MattermostConfig(
            url=_resolve_env(mattermost_data.get("url", "")),
            token=_resolve_env(mattermost_data.get("token", "")),
            port=mattermost_data.get("port", 8065),
        ),
        discord=DiscordConfig(
            token=_resolve_env(discord_data.get("token", "")),
            guild_id=discord_data.get("guild_id", ""),
        ),
        channels=slack.get("channels", []),
        persona=PersonaConfig(
            display_name=persona_data.get("display_name", ""),
            description=persona_data.get("description", ""),
            personality=persona_data.get("personality", ""),
            response_prefix=persona_data.get("response_prefix", ""),
        ),
        project=ProjectConfig(
            repo_path=project_data.get("repo_path", ""),
            github_repo=project_data.get("github_repo", ""),
            github_token_var=project_data.get("github_token_var", ""),
        ),
        airflow=AirflowConfig(dag_prefix=airflow_data.get("dag_prefix", "")),
        tools_enabled=tools_data.get("enabled", []),
        claude=ClaudeConfig(
            provider=claude_data.get("provider", "claude"),
            model=claude_data.get("model", "sonnet"),
            max_turns=claude_data.get("max_turns", 10),
            full_agent=claude_data.get("full_agent", False),
        ),
        memory=MemoryConfig(
            context_window=memory_data.get("context_window", 20),
            auto_extract=memory_data.get("auto_extract", True),
        ),
        security=SecurityConfig(
            allowed_channels=security_data.get("allowed_channels", []),
            allowed_users=security_data.get("allowed_users", []),
            dangerous_tools=security_data.get("dangerous_tools", []),
            mention_only=security_data.get("mention_only", False),
        ),
    )
    _validate_bot_config(config)
    return config


def _as_dict(value) -> dict:
    if isinstance(value, dict):
        return value
    if not value:
        return {}
    if isinstance(value, str):
        return json.loads(value)
    return dict(value)


def _as_list(value) -> list:
    if isinstance(value, list):
        return value
    if not value:
        return []
    if isinstance(value, str):
        return json.loads(value)
    return list(value)


def load_bot_from_db_row(row) -> BotConfig:
    """Load a single bot configuration from a database row.

    Supports both old 12-column schema and new 15-column schema that
    adds platform, discord, and mattermost columns. Missing columns
    fall back to safe defaults for backward compatibility.
    """
    bot_id = row[0]
    name = row[1]
    slack_app_token = row[2]
    slack_bot_token = row[3]
    channels = row[4]
    persona_data = row[5]
    project_data = row[6]
    airflow_data = row[7]
    tools_data = row[8]
    claude_data = row[9]
    memory_data = row[10]
    security_data = row[11]
    platform = row[12] if len(row) > 12 else "slack"
    discord_data = row[13] if len(row) > 13 else {}
    mattermost_data = row[14] if len(row) > 14 else {}

    persona_data = _as_dict(persona_data)
    project_data = _as_dict(project_data)
    airflow_data = _as_dict(airflow_data)
    tools_data = _as_dict(tools_data)
    claude_data = _as_dict(claude_data)
    memory_data = _as_dict(memory_data)
    security_data = _as_dict(security_data)
    discord_data = _as_dict(discord_data)
    mattermost_data = _as_dict(mattermost_data)

    return BotConfig(
        id=bot_id,
        name=name,
        slack_app_token=_resolve_env(slack_app_token or ""),
        slack_bot_token=_resolve_env(slack_bot_token or ""),
        platform=platform or "slack",
        discord=DiscordConfig(
            token=_resolve_env(discord_data.get("token", "")),
            guild_id=discord_data.get("guild_id", ""),
        ),
        mattermost=MattermostConfig(
            url=_resolve_env(mattermost_data.get("url", "")),
            token=_resolve_env(mattermost_data.get("token", "")),
            port=mattermost_data.get("port", 8065),
        ),
        channels=_as_list(channels),
        persona=PersonaConfig(
            display_name=persona_data.get("display_name", name or bot_id or ""),
            description=persona_data.get("description", ""),
            personality=persona_data.get("personality", ""),
            response_prefix=persona_data.get("response_prefix", ""),
        ),
        project=ProjectConfig(
            repo_path=project_data.get("repo_path", ""),
            github_repo=project_data.get("github_repo", ""),
            github_token_var=project_data.get("github_token_var", ""),
        ),
        airflow=AirflowConfig(dag_prefix=airflow_data.get("dag_prefix", "")),
        tools_enabled=_as_list(tools_data.get("enabled", [])),
        claude=ClaudeConfig(
            provider=claude_data.get("provider", "claude"),
            model=claude_data.get("model", "sonnet"),
            max_turns=claude_data.get("max_turns", 10),
            full_agent=claude_data.get("full_agent", False),
        ),
        memory=MemoryConfig(
            context_window=memory_data.get("context_window", 20),
            auto_extract=memory_data.get("auto_extract", True),
        ),
        security=SecurityConfig(
            allowed_channels=_as_list(security_data.get("allowed_channels", [])),
            allowed_users=_as_list(security_data.get("allowed_users", [])),
            dangerous_tools=_as_list(security_data.get("dangerous_tools", [])),
        ),
    )


def load_all_bots_from_db() -> list[BotConfig]:
    """Load all active bot configurations from PostgreSQL."""
    dsn = os.environ.get("DATABASE_DSN", "")
    if not dsn:
        return []

    conn = psycopg2.connect(dsn)
    try:
        with conn.cursor() as cur:
            cur.execute(
                """
                SELECT
                    id,
                    name,
                    slack_app_token,
                    slack_bot_token,
                    channels,
                    persona,
                    project,
                    airflow,
                    tools,
                    claude,
                    memory,
                    security,
                    platform,
                    discord,
                    mattermost
                FROM bots
                WHERE is_active = true
                ORDER BY created_at, id
                """
            )
            rows = cur.fetchall()
        configs = []
        for row in rows:
            try:
                configs.append(load_bot_from_db_row(row))
            except (ValueError, Exception) as e:
                print(f"[WARNING] Skipping bot {row[0]}: {e}")
        return configs
    finally:
        conn.close()


def load_all_bots(bots_dir: str = "/app/bots") -> list[BotConfig]:
    """Load all bot configurations, preferring PostgreSQL over YAML bootstrap."""
    configs_by_id: dict[str, BotConfig] = {}

    try:
        for config in load_all_bots_from_db():
            configs_by_id[config.id] = config
    except Exception as e:
        print(f"[WARNING] Failed to load bot configs from database: {e}")

    bots_path = Path(bots_dir)
    if bots_path.exists():
        for yaml_file in sorted(bots_path.glob("*.yaml")):
            try:
                config = load_bot_from_yaml(yaml_file)
                if config.id not in configs_by_id:
                    configs_by_id[config.id] = config
            except Exception as e:
                print(f"[WARNING] Failed to load bot config {yaml_file}: {e}")
    return list(configs_by_id.values())

File: go/internal/config/loader.go
package config

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"

	"github.com/jackc/pgx/v5"
	"gopkg.in/yaml.v3"
)

// resolveEnv replaces ${ENV_VAR} references in value with the corresponding
// environment variable. Returns an empty string if the variable is not set.
func resolveEnv(value string) string {
	if strings.HasPrefix(value, "${") && strings.HasSuffix(value, "}") {
		key := value[2 : len(value)-1]
		return os.Getenv(key)
	}
	return value
}

// yamlBotFile is the raw YAML structure as read from disk.
// Fields are mapped to BotConfig after env resolution.
type yamlBotFile struct {
	Name        string            `yaml:"name"`
	Platform    string            `yaml:"platform"`
	Slack       map[string]any    `yaml:"slack"`
	Mattermost  map[string]any    `yaml:"mattermost"`
	Discord     map[string]any    `yaml:"discord"`
	Channels    []string          `yaml:"channels"`
	Persona     map[string]any    `yaml:"persona"`
	Project     map[string]any    `yaml:"project"`
	Airflow     map[string]any    `yaml:"airflow"`
	Tools       map[string]any    `yaml:"tools"`
	Claude      map[string]any    `yaml:"claude"`
	Memory      map[string]any    `yaml:"memory"`
	Security    map[string]any    `yaml:"security"`
	// Credentials is the platform-neutral key-value credential store.
	// Values that begin with "${" are resolved as environment variables.
	Credentials map[string]string `yaml:"credentials"`
}

func stringField(m map[string]any, key string) string {
	if m == nil {
		return ""
	}
	v, _ := m[key].(string)
	return v
}

func intField(m map[string]any, key string, def int) int {
	if m == nil {
		return def
	}
	switch v := m[key].(type) {
	case int:
		return v
	case float64:
		return int(v)
	}
	return def
}

func boolField(m map[string]any, key string, def bool) bool {
	if m == nil {
		return def
	}
	v, ok := m[key].(bool)
	if !ok {
		return def
	}
	return v
}

func stringSliceField(m map[string]any, key string) []string {
	if m == nil {
		return nil
	}
	raw, _ := m[key].([]any)
	out := make([]string, 0, len(raw))
	for _, item := range raw {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

// LoadBotFromYAML reads a single bot configuration from a YAML file.
func LoadBotFromYAML(path string) (*BotConfig, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", path, err)
	}

	var raw yamlBotFile
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parse %s: %w", path, err)
	}

	enabledTools := stringSliceField(raw.Tools, "enabled")

	cfg := &BotConfig{
		ID:            raw.Name,
		Name:          raw.Name,
		Platform:      raw.Platform,
		SlackAppToken: resolveEnv(stringField(raw.Slack, "app_token")),
		SlackBotToken: resolveEnv(stringField(raw.Slack, "bot_token")),
		Mattermost: MattermostConfig{
			URL:   resolveEnv(stringField(raw.Mattermost, "url")),
			Token: resolveEnv(stringField(raw.Mattermost, "token")),
			Port:  intField(raw.Mattermost, "port", 8065),
		},
		Discord: DiscordConfig{
			Token:   resolveEnv(stringField(raw.Discord, "token")),
			GuildID: stringField(raw.Discord, "guild_id"),
		},
		Channels: raw.Channels,
		Persona: PersonaConfig{
			DisplayName:    stringField(raw.Persona, "display_name"),
			Description:    stringField(raw.Persona, "description"),
			Personality:    stringField(raw.Persona, "personality"),
			ResponsePrefix: stringField(raw.Persona, "response_prefix"),
		},
		Project: ProjectConfig{
			RepoPath:       stringField(raw.Project, "repo_path"),
			GithubRepo:     stringField(raw.Project, "github_repo"),
			GithubTokenVar: stringField(raw.Project, "github_token_var"),
		},
		Airflow: AirflowConfig{
			DagPrefix: stringField(raw.Airflow, "dag_prefix"),
		},
		ToolsEnabled: enabledTools,
		Claude: ClaudeConfig{
			Provider:  stringField(raw.Claude, "provider"),
			Model:     stringField(raw.Claude, "model"),
			MaxTurns:  intField(raw.Claude, "max_turns", 5),
			FullAgent: boolField(raw.Claude, "full_agent", false),
		},
		Memory: MemoryConfig{
			ContextWindow: intField(raw.Memory, "context_window", 20),
			AutoExtract:   boolField(raw.Memory, "auto_extract", true),
		},
		Security: SecurityConfig{
			AllowedChannels: stringSliceField(raw.Security, "allowed_channels"),
			AllowedUsers:    stringSliceField(raw.Security, "allowed_users"),
			DangerousTools:  stringSliceField(raw.Security, "dangerous_tools"),
			MentionOnly:     boolField(raw.Security, "mention_only", false),
		},
	}
	cfg.defaults()

	if err := validateBotConfig(cfg); err != nil {
		return nil, fmt.Errorf("validate %s: %w", path, err)
	}
	return cfg, nil
}

// LoadAllBotsFromDB loads all active bot configurations from PostgreSQL.
// It returns an empty slice (not an error) when DATABASE_DSN is unset.
func LoadAllBotsFromDB(ctx context.Context, dsn string) ([]*BotConfig, error) {
	if dsn == "" {
		return nil, nil
	}

	conn, err := pgx.Connect(ctx, dsn)
	if err != nil {
		return nil, fmt.Errorf("connect to database: %w", err)
	}
	defer func() { _ = conn.Close(ctx) }()

	rows, err := conn.Query(ctx, `
		SELECT
			id, name, slack_app_token, slack_bot_token,
			channels, persona, project, airflow, tools,
			claude, memory, security,
			platform, discord, mattermost,
			anthropic_api_key, openai_api_key, gemini_api_key,
			credentials
		FROM bots
		WHERE is_active = true
		ORDER BY created_at, id
	`)
	if err != nil {
		return nil, fmt.Errorf("query bots: %w", err)
	}
	defer rows.Close()

	var configs []*BotConfig
	for rows.Next() {
		cfg, err := scanBotRow(rows)
		if err != nil {
			slog.Warn("skipping bot row", "error", err)
			continue
		}
		cfg.defaults()
		configs = append(configs, cfg)
	}
	return configs, rows.Err()
}

// scanBotRow scans a single database row into a BotConfig.
func scanBotRow(rows pgx.Rows) (*BotConfig, error) {
	var (
		id, name, slackApp, slackBot string
		channels                     []string
		persona, project, airflow    map[string]any
		tools, claude, memory        map[string]any
		security                     map[string]any
		platform                     string
		discord, mattermost          map[string]any
		anthropicKey, openaiKey, geminiKey *string // nullable per-bot LLM keys
		credentialsRaw               []byte
	)

	err := rows.Scan(
		&id, &name, &slackApp, &slackBot,
		&channels, &persona, &project, &airflow, &tools,
		&claude, &memory, &security,
		&platform, &discord, &mattermost,
		&anthropicKey, &openaiKey, &geminiKey,
		&credentialsRaw,
	)
	if err != nil {
		return nil, fmt.Errorf("scan row: %w", err)
	}

	enabledTools := stringSliceField(tools, "enabled")

	cfg := &BotConfig{
		ID:            id,
		Name:          name,
		SlackAppToken: resolveEnv(slackApp),
		SlackBotToken: resolveEnv(slackBot),
		Platform:      platform,
		Channels:      channels,
		Mattermost: MattermostConfig{
			URL:   resolveEnv(stringField(mattermost, "url")),
			Token: resolveEnv(stringField(mattermost, "token")),
			Port:  intField(mattermost, "port", 8065),
		},
		Discord: DiscordConfig{
			Token:   resolveEnv(stringField(discord, "token")),
			GuildID: stringField(discord, "guild_id"),
		},
		Persona: PersonaConfig{
			DisplayName:    stringField(persona, "display_name"),
			Description:    stringField(persona, "description"),
			Personality:    stringField(persona, "personality"),
			ResponsePrefix: stringField(persona, "response_prefix"),
		},
		Project: ProjectConfig{
			RepoPath:       stringField(project, "repo_path"),
			GithubRepo:     stringField(project, "github_repo"),
			GithubTokenVar: stringField(project, "github_token_var"),
		},
		Airflow: AirflowConfig{
			DagPrefix: stringField(airflow, "dag_prefix"),
		},
		ToolsEnabled: enabledTools,
		Claude: ClaudeConfig{
			Provider:  stringField(claude, "provider"),
			Model:     stringField(claude, "model"),
			MaxTurns:  intField(claude, "max_turns", 5),
			FullAgent: boolField(claude, "full_agent", false),
		},
		Memory: MemoryConfig{
			ContextWindow: intField(memory, "context_window", 20),
			AutoExtract:   boolField(memory, "auto_extract", true),
		},
		Security: SecurityConfig{
			AllowedChannels: stringSliceField(security, "allowed_channels"),
			AllowedUsers:    stringSliceField(security, "allowed_users"),
			DangerousTools:  stringSliceField(security, "dangerous_tools"),
			MentionOnly:     boolField(security, "mention_only", false),
		},
	}

	if anthropicKey != nil {
		cfg.LLMKeys.AnthropicKey = *anthropicKey
	}
	if openaiKey != nil {
		cfg.LLMKeys.OpenAIKey = *openaiKey
	}
	if geminiKey != nil {
		cfg.LLMKeys.GeminiKey = *geminiKey
	}

	if len(credentialsRaw) > 0 && string(credentialsRaw) != "{}" {
		if err := json.Unmarshal(credentialsRaw, &cfg.Credentials); err != nil {
			slog.Warn("failed to parse credentials", "bot", cfg.ID, "error", err)
		}
	}

	return cfg, nil
}

// LoadAllBots loads all bot configurations, preferring the database over YAML files.
// YAML files in botsDir act as a bootstrap fallback: a bot loaded from the database
// will not be overridden by a YAML file with the same ID.
func LoadAllBots(ctx context.Context, botsDir, dsn string) ([]*BotConfig, error) {
	byID := make(map[string]*BotConfig)

	dbConfigs, err := LoadAllBotsFromDB(ctx, dsn)
	if err != nil {
		slog.Warn("failed to load bots from database, falling back to YAML", "error", err)
	}
	for _, cfg := range dbConfigs {
		byID[cfg.ID] = cfg
	}

	entries, err := filepath.Glob(filepath.Join(botsDir, "*.yaml"))
	if err != nil {
		return nil, fmt.Errorf("glob bots dir: %w", err)
	}
	for _, path := range entries {
		cfg, err := LoadBotFromYAML(path)
		if err != nil {
			slog.Warn("skipping bot YAML", "path", path, "error", err)
			continue
		}
		if _, exists := byID[cfg.ID]; !exists {
			byID[cfg.ID] = cfg
		}
	}

	configs := make([]*BotConfig, 0, len(byID))
	for _, cfg := range byID {
		configs = append(configs, cfg)
	}
	return configs, nil
}

// validateBotConfig checks that the required platform-specific credentials are present.
func validateBotConfig(cfg *BotConfig) error {
	if cfg.Platform == "" {
		return fmt.Errorf("bot %q: platform must be explicitly set", cfg.ID)
	}
	switch cfg.Platform {
	case "slack":
		appToken := cfg.SlackAppToken
		botToken := cfg.SlackBotToken
		if appToken == "" {
			appToken = cfg.Credentials["app_token"]
		}
		if botToken == "" {
			botToken = cfg.Credentials["bot_token"]
		}
		if appToken == "" || botToken == "" {
			return fmt.Errorf("bot %q: slack requires app_token and bot_token (via credentials or legacy fields)", cfg.ID)
		}
	case "mattermost":
		url := cfg.Mattermost.URL
		token := cfg.Mattermost.Token
		if url == "" {
			url = cfg.Credentials["url"]
		}
		if token == "" {
			token = cfg.Credentials["token"]
		}
		if url == "" || token == "" {
			return fmt.Errorf("bot %q: mattermost requires url and token", cfg.ID)
		}
	case "discord":
		token := cfg.Discord.Token
		if token == "" {
			token = cfg.Credentials["token"]
		}
		if token == "" {
			return fmt.Errorf("bot %q: discord requires token", cfg.ID)
		}
	default:
		return fmt.Errorf("bot %q has unsupported platform %q; supported: slack, mattermost, discord",
			cfg.ID, cfg.Platform)
	}
	return nil
}
